backend/planty/dashboard/views.py

```diff
@@ -290,3 +295,234 @@ def post(self, request: Request):
+
+    def post(self, request: Request):
+        user: User = request.user
+        serializer = InstructionCreateSerializer(data=request.data)
+
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        data = serializer.validated_data
+
+        instruction: Instruction = Instruction.objects.create(
+            id=uuid4(),
+            name=name,
+
+            user=user,
+            species=data['species'],
+
+            watering=data['watering'],
+            insolation=data['insolation'],
+            fertilizing=data['fertilizing']
+        )
+        instruction.save()
+
+        return Response(status=status.HTTP_201_CREATED)
+
+    def put(self, request: Request):
+        user: User = request.user
+        serializer = InstructionSelectSerializer(data=request.data)
+
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        data = serializer.validated_data
+
+        try:
+            instruction: Instruction = Instruction.objects.get(pk=data['id'])
+        except Model.DoesNotExist:
+            return Response(data={
+                'id': ['Plant with the given ID does not exist']
+            }, status=status.HTTP_404_NOT_FOUND)
+
+        if instruction.user != user:
+            return Response(status=status.HTTP_403_FORBIDDEN)
+
+        if instruction.public == 'True':
+            return Response(data={
+                'id:': ['Instruction with this ID is public. Cannot modify public instructions.']
+            }, status=status.HTTP_404_NOT_FOUND)
+
+        instruction_json = {
+            'id': str(instruction.id),
+            'name': instruction.name,
+            'species': instruction.species,
+
+            'watering': instruction.watering,
+            'insolation': instruction.insolation,
+            'fertilizing': instruction.fertilizing,
+
+            'public' : instruction.public
+        }
+
+        return Response(instruction_json, status=status.HTTP_200_OK)
+
+    def delete(self, request: Request):
+        user: User = request.user
+        serializer = InstructionDeleteSerializer(data=request.data)
+
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        data = serializer.validated_data
+
+        try:
+            instruction: Instruction = Instruction.objects.get(pk=data['id'])
+        except Model.DoesNotExist:
+            return Response(data={
+                'id': ['Instruction with the given ID does not exist']
+            }, status=status.HTTP_204_NOT_FOUND)
+
+        if instruction.user == user:
+            plant.delete()
+        else:
+            return Response(status=status.HTTP_403_FORBIDDEN)
+
+        return Response(status=status.HTTP_200_OK)
+
+
+class SelectInstruction(APIView):
```

Comment:
Trochę mi zajęło zanim zrozumiałem co się dzieje w tych dwóch ostatnich klasach, wyglądają prawie identycznie.
Myślę że to rozwiązanie nie jest dobre.
Jeśli dobrze rozumiem to potrzebujemy dwóch endpointów:
- `GET /instructions/public` - zwraca publiczne instrukcje
- Jakiś endpoint do publikowania. Ja bym zrobił `POST /instructions/public`

Już pomijając sens tego endpointa, nie rozumiem też dlaczego GET. GET nie powinien tworzyć nowych zasobów ani niczego modyfikować. PUT powinien tylko modyfikować.

Jeśli chodzi o inkrementowanie licznika popularności to powinno to się odbywać przy tworzeniu/modyfikacji rośliny. Nie można liczyć na to że client sam wywołałby taki endpoint. Ktoś może sobie zamiast z apki webowej korzystać z API bezpośrednio i nie wywołać tego endpointa wcale, albo wywołać go 9999 razy dla jednej instrukcji.

Reply:
idea była taka, żeby w `GET 'instructions/select'` tworzyć i jednocześnie zwracać kopię z jakiejś upublicznionej instrukcji, którą wybierze użytkownik z nowym `id` i `public=False`(pomyliłem się tutaj i wpisałem `public=True`), żeby nowo utworzona instrukcja nie była w tych publicznych

natomiast `PUT 'instructions/share'` miał tworzyć kopię własnej instrukcji z nowym `id` i `public=True`, żeby ta instrukcja wylądowała w publicznych i mogła się znaleźć w rankingu

Aby wyeliminować efekt klikania własnych instrukcji, żeby znaleźć się wyżej w rankingu wybrań i łatwego inkrementowania licznika, można rozważyć filtrowanie jedynie publicznych instrukcji, ale tylko innych użytkowników